autopush/web/registration.py

```diff
@@ -0,0 +1,311 @@
+import json
+import re
+import time
+import uuid
+
+from autopush.exceptions import InvalidRequest
+from autopush.web.base import threaded_validate, BaseWebHandler, PREF_SCHEME, \
+    AUTH_SCHEMES
+from autopush.db import (
+    generate_last_connect,
+    hasher,
+)
+from autopush.utils import generate_hash, ms_time
+from boto.dynamodb2.exceptions import ItemNotFound
+from cryptography.hazmat.primitives import constant_time
+from marshmallow import Schema, fields, pre_load, validates_schema
+from twisted.internet.defer import Deferred
+from twisted.internet.threads import deferToThread
+
+
+class RegistrationSchema(Schema):
+    uaid = fields.Str(allow_none=True)
+    chid = fields.Str(allow_none=True)
+    body = fields.Dict(allow_none=True)
+    router_type = fields.Str()
+    router_token = fields.Str()
+    params = fields.Dict()
+    auth = fields.Str(allow_none=True)
+    vapid_info = fields.Dict(allow_none=True)
+
+    @pre_load
+    def extract_data(self, req):
+        params = {}
+        if req['body']:
+            try:
+                params = json.loads(req['body'])
+            except ValueError:
+                raise InvalidRequest("Invalid Request body",
+                                     status_code=401,
+                                     errno=108)
+        uaid = req['path_kwargs'].get('uaid')
+        chid = req['path_kwargs'].get('chid', params.get("channelID"))
+        if uaid:
+            try:
+                uuid.UUID(uaid)
+            except (ValueError, TypeError):
+                raise InvalidRequest("Invalid Request UAID",
+                                     status_code=401, errno=109)
+        if chid:
+            try:
+                uuid.UUID(chid)
+            except (ValueError, TypeError):
+                raise InvalidRequest("Invalid Request Channel_id",
+                                     status_code=410, errno=106)
+        return dict(
+            auth=req.get('headers', {}).get("Authorization"),
+            params=params,
+            router_type=req['path_kwargs'].get('router_type'),
+            router_token=req['path_kwargs'].get('router_token'),
+            uaid=uaid,
+            chid=chid,
```

Comment:
No. Delete is either the UAID + CHID or just the UAID for a full deregistration.

I suppose I can add special validation handing for the various message types, but I was thinking that it would be clearer to have that in the handlers, where we have to do the logic anyway.